sycl/doc/extensions/proposed/sycl_ext_oneapi_non_uniform_groups.asciidoc

```diff
@@ -0,0 +1,1161 @@
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+
+== Overview
+
+Many modern hardware architectures support flexible sub-divisions of
+work-groups and sub-groups to support fine-grained work scheduling. A common
+use-case for such flexibility is communication and synchronization between
+work-items in divergent control flow.
+
+This proposal introduces new classes to represent sub-divisions of SYCL's
+built-in group types, traits for detecting these classes, and free functions
+for creating new instances of these classes.
+
+These new classes can be used as arguments to group functions, group algorithms,
+and custom functions to convey exactly which work-items an operation is expecting,
+simplifying interfaces and greatly reducing the amount of delicate wording
+needed in function documentation.
+
+NOTE: The first version of this extension only supports partitioning of
+sub-groups. It is expected that in the future, these functions will be expanded
+to also allow partitioning of root-groups, work-groups and user-constructed
+groups.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_NON_UNIFORM_GROUPS` to one of the values defined in the
+table below.  Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== Control Flow
+
+The SYCL specification defines
+link:https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#control-flow[control
+flow] as below:
+
+> When all work-items in a group are executing the same sequence of statements,
+> they are said to be executing under _converged_ control flow. Control flow
+> _diverges_ when different work-items in a group execute a different sequence of
+> statements, typically as a result of evaluating conditions differently (e.g.
+> in selection statements or loops).
+
+This extension introduces some new terminology to describe other kinds of
+control flow, to simplify the description of the behavior for new group types.
+
+A work-item is said to have _active_ control flow for a given point of
+execution if it follows the sequence of statements currently being executed.
```

Comment:
You're right, this is unclear.  I think you're also right that there aren't currently any dependencies on "active control flow", so we can probably remove this definition.

Do you think we need to define what "active" means, instead?

Reply:
The only place the word "active" occurs in this extension is in the definition of `get_masked_sub_group`.  I'm not really sure what we mean by "active" when we define that function.  I thought the function just created a group containing the work-items identified by the mask, in which case the word "active" doesn't mean anything.  Is there some extra requirement that a work-item must both be in the mask and also "active" in order to be part of the group?

Note that the word "active" also shows up in the examples as part of a variable name (e.g. `active_group`).  This seems fine if we decide to remove the formal term "active control flow" from the spec.  However, if "active" has some formal meaning, we should only name a variable "active" if it agrees with that formal meaning.

Reply:
I think you're right.  "active" is being used redundantly here, because the important thing is whether or not a work-item has a corresponding bit set in the mask.  Because of the way we've defined other functions, we don't need this definition -- when we say something like "all work-items in the group must encounter this function", it already means "all work-items with a corresponding bit set in the mask must encounter this function".

I've made this change in 7ad1cd274.